File: examples/Perl/EX7/telephone.pl
#!/usr/bin/env perl
# -*- tab-width: 4; -*-

#
# The contents of this file are subject to the Mozilla Public
# License Version 1.1 (the "License"); you may not use this file
# except in compliance with the License. You may obtain a copy of
# the License at http://www.mozilla.org/MPL/
#
# Software distributed under the License is distributed on an "AS
# IS" basis, WITHOUT WARRANTY OF ANY KIND, either express or
# implied. See the License for the specific language governing
# rights and limitations under the License.
#
# The Original Code is State Machine Compiler (SMC).
#
# The Initial Developer of the Original Code is Charles W. Rapp.
# Portions created by Charles W. Rapp are
# Copyright (C) 2000 - 2003 Charles W. Rapp.
# All Rights Reserved.
#
# Contributor(s):
#       Port to Perl by Francois Perrad, francois.perrad@gadz.org
#
# Name
#  telephone.pl
#
# Description
#  A simulation of an old fashioned touch-tone telephone.
#
# RCS ID
# $Id: telephone.pl,v 1.3 2009/04/22 19:07:03 fperrad Exp $
#
# CHANGE LOG
# $Log: telephone.pl,v $
# Revision 1.3  2009/04/22 19:07:03  fperrad
# Added enterStartState method
#
# Revision 1.2  2008/02/04 12:40:11  fperrad
# some Perl Best Practices
#
# Revision 1.1  2005/06/16 18:04:15  fperrad
# Added Perl examples 1 - 4 and 7.
#
#

use strict;
use warnings;

use Tk;

package Telephone;

use POSIX qw(strftime);
use Telephone_sm;

use constant LONG_DISTANCE => 1;
use constant LOCAL => 2;
use constant EMERGENCY => 3;

use constant NYC_TEMP => 4;
use constant TIME => 5;
use constant DEPOSIT_MONEY => 6;
use constant LINE_BUSY => 7;
use constant INVALID_NUMBER => 8;

use constant SEC_PER_MINUTE => 60;

sub new {
    my $proto = shift;
    my $class = ref($proto) || $proto;
    my $self = {
            _areaCode => q{},
            _exchange => q{},
            _local => q{},
            _display => q{},
            _receiverButton => undef,
            _timerMap => {
                ClockTimer => -1,
                OffHookTimer => -1,
                LoopTimer => -1,
                RingTimer => -1,
            },
            _timerAudioID => -1,
    };
    bless($self, $class);

    $self->_loadUI();

    # Create the state machine to drive this object.
    $self->{_fsm} = new smc_ex7::Telephone_sm($self);

    # DEBUG
    #$self->{_fsm}->setDebugFlag(1);

    return $self;
}

# Create the user interface but don't display it yet.
sub _loadUI {
    my $self = shift;
    my $mw = new MainWindow();
    $self->{_mw} = $mw;
    $mw->title('Telephone demo');

    my $frameDisplay = $mw->Frame()->pack(
            -side => 'top',
            -fill => 'both',
    );
    # Create the read-only phone number display.
    $self->{_numberDisplay} = $frameDisplay->Label(
            -width => 30,
            -bg => 'white',
            -relief => 'sunken',
            -padx => 5,
            -pady => 5,
    )->pack(
            -side => 'top',
            -padx => 5,
            -pady => 5,
    );

    my $frameHook = $mw->Frame()->pack(
            -side => 'top',
            -expand => 1,
    );
    # Create the off-hook/on-hook button.
    $self->{_receiverButton} = $frameHook->Button(
            -text => 'Pick up receiver',
            -state => 'normal',
            -command => sub {
                my $text = $self->{_receiverButton}->cget(-text);
                if ($text eq 'Pick up receiver') {
                    $self->{_fsm}->OffHook();
                }
                elsif ($text eq 'Put down receiver') {
                    $self->{_fsm}->OnHook();
                }
                else {
                    warn "Unknown receiver ($text).\n";
                }
            },
            -padx => 10,
            -pady => 5,
            -bd => 3,
    )->pack(
            -side => 'top',
            -padx => 5,
            -pady => 5,
    );

    my $frameDial = $mw->Frame()->pack(
            -side => 'top',
            -fill => 'both',
            -padx => 5,
            -pady => 5,
    );
    # Create the dialing buttons.
    my @w;
    foreach my $digit (1..9, '*', 0, '#') {
        push @w, $frameDial->Button(
                    -text => $digit,
                    -height => 2,
                    -width => 3,
                    -command => sub {
                        $self->{_fsm}->Digit($digit);
                    },
            );
    }
    $w[0]->grid($w[1], $w[2],
            -padx => 1,
            -pady => 1,
    );
    $w[3]->grid($w[4], $w[5],
            -padx => 1,
            -pady => 1,
    );
    $w[6]->grid($w[7], $w[8],
            -padx => 1,
            -pady => 1,
    );
    $w[9]->grid($w[10], $w[11],
            -padx => 1,
            -pady => 1,
    );

    my $frameStatus = $mw->Frame()->pack(
            -side => 'top',
            -fill => 'both',
    );
    $self->{_soundDisplay} = $frameStatus->Label(
            -relief => 'groove',
    )->pack(
            -side => 'top',
            -fill => 'x',
            -expand => 1,
            -anchor => 's',
    );

    # Cntl-C stops the demo as well.
    $mw->bind('<Control-c>',
        [sub { exit(0); }]
    );
}

sub Start {
    my $self = shift;
    $self->{_fsm}->enterStartState();
}

#-----------------------------------------------------------
# State Machine Actions.
#

# Return the current area code.
sub getAreaCode {
    my $self = shift;
    return $self->{_areaCode};
}

# Return the exchange.
sub getExchange {
    my $self = shift;
    return $self->{_exchange};
}

# Return the local number.
sub getLocal {
    my $self = shift;
    return $self->{_local};
}

sub routeCall {
    my $self = shift;
    my ($callType, $areaCode, $exchange, $local) = @_;
    my $route;

    if ($callType == EMERGENCY) {
        $route = EMERGENCY;
    }
    elsif ($callType == LONG_DISTANCE
        && $areaCode eq '1212'
        && $exchange eq '555'
        && $local eq '1234') {
        $route = NYC_TEMP;
    }
    elsif ($exchange eq '555') {
        if ($local eq '1212') {
            $route = TIME;
        }
        else {
            $route = LINE_BUSY;
        }
    }
    elsif ($callType == LOCAL) {
        $route = DEPOSIT_MONEY;
    }
    else {
        $route = INVALID_NUMBER;
    }

    # Call routing needs to be done asynchronouzly in order to
    # avoid issuing a transition within a transition.
    $self->{_mw}->after(50, sub { $self->_callRoute($route); });
}

sub startTimer {
    my $self = shift;
    my ($name, $delay) = @_;
    if ($name eq 'ClockTimer') {
        $self->{_timerMap}->{$name} =
            $self->{_mw}->after($delay, sub { $self->{_fsm}->ClockTimer(); });
    }
    elsif ($name eq 'OffHookTimer') {
        $self->{_timerMap}->{$name} =
            $self->{_mw}->after($delay, sub { $self->{_fsm}->OffHookTimer(); });
    }
    elsif ($name eq 'LoopTimer') {
        $self->{_timerMap}->{$name} =
            $self->{_mw}->after($delay, sub { $self->{_fsm}->LoopTimer(); });
    }
    elsif ($name eq 'RingTimer') {
        $self->{_timerMap}->{$name} =
            $self->{_mw}->after($delay, sub { $self->{_fsm}->RingTimer(); });
    }
}

sub resetTimer {
    my $self = shift;
    my ($name, $delay) = @_;
    if ($self->{_timerMap}->{$name} >= 0) {
        $self->{_mw}->afterCancel($self->{_timerMap}->{$name});
        $self->{_timerMap}->{$name} = -1;
        $self->startTimer($name, $delay);
    }
}

sub stopTimer {
    my $self = shift;
    my ($name) = @_;
    if ($self->{_timerMap}->{$name} >= 0) {
        $self->{_mw}->afterCancel($self->{_timerMap}->{$name});
        $self->{_timerMap}->{$name} = -1;
    }
}

sub play {
    my $self = shift;
    my ($name, $delay) = @_;
    if ($self->{_timerAudioID} >= 0) {
        $self->{_mw}->afterCancel($self->{_timerAudioID});
        $self->{_timerAudioID} = -1;
    }
    $self->{_soundDisplay}->configure(-text => $name);
    $self->{_timerAudioID} = $self->{_mw}->after($delay, sub {
        $self->{_soundDisplay}->configure(-text => q{});
        $self->{_timerAudioID} = -1;
    });
}

sub playTT {
    my $self = shift;
    my ($name) = @_;
    $self->play($name, 400);
}

sub loop {
    my $self = shift;
    my ($name) = @_;
    if ($self->{_timerAudioID} >= 0) {
        $self->{_mw}->afterCancel($self->{_timerAudioID});
        $self->{_timerAudioID} = -1;
    }
    $self->{_soundDisplay}->configure(-text => $name . ' ...');
}

sub stopLoop {
    my $self = shift;
    $self->{_soundDisplay}->configure(-text => q{});
}

sub stopPlayback {
    my $self = shift;
    if ($self->{_timerAudioID} >= 0) {
        $self->{_mw}->afterCancel($self->{_timerAudioID});
        $self->{_timerAudioID} = -1;
    }
    $self->{_soundDisplay}->configure(-text => q{});
}

sub playEmergency {
    my $self = shift;
    $self->play('911', 5000);
}

sub playNYCTemp {
    my $self = shift;
    $self->play('NYC_temp', 2000);
}

sub playDepositMoney {
    my $self = shift;
    $self->play('50_cents_please', 2000);
}

sub playTime {
    my $self = shift;
    $self->play('the_time_is ???', 2000);
}

sub playInvalidNumber {
    my $self = shift;
    $self->play('you_dialed ### could_not_be_completed', 2000);
}

sub getType {
    my $self = shift;
    return $self->{_callType};
}

sub setType {
    my $self = shift;
    my ($type) = @_;
    $self->{_callType} = $type;
}

sub saveAreaCode {
    my $self = shift;
    my ($n) = @_;
    $self->{_areaCode} .= $n;
    $self->addDisplay($n);
}

sub saveExchange {
    my $self = shift;
    my ($n) = @_;
    $self->{_exchange} .= $n;
    $self->addDisplay($n);
}

sub saveLocal {
    my $self = shift;
    my ($n) = @_;
    $self->{_local} .= $n;
    $self->addDisplay($n);
}

sub addDisplay {
    my $self = shift;
    my ($character) = @_;
    $self->{_display} .= $character;
    $self->{_numberDisplay}->configure(-text => $self->{_display});
}

sub clearDisplay {
    my $self = shift;

    #Clear the internal data store.
    $self->{_display} = q{};
    $self->{_areaCode} = q{};
    $self->{_local} = q{};
    $self->{_exchange} = q{};

    # Put up the current time and date on the display.
    $self->{_numberDisplay}->configure(-text => q{});
}

sub startClockTimer {
    my $self = shift;
    my $currentTime = time();
    my $timeRemaining = SEC_PER_MINUTE - ($currentTime % SEC_PER_MINUTE);

    # Figure out how long until the top of the minute
    # and set the timer for that amount.
    $self->startTimer('ClockTimer', $timeRemaining * 1000);
}

sub updateClock {
    my $self = shift;
    my $text = strftime('%H:%M  %b %d, %Y', localtime(time()));
    $self->{_numberDisplay}->configure(-text => $text);
}

sub setReceiver {
    my $self = shift;
    my ($text) = @_;
        if (defined $self->{_receiverButton}) {
        $self->{_receiverButton}->configure(-text => $text);
    }
}

sub _callRoute {
    my $self = shift;
    my ($route) = @_;
    if ($route == EMERGENCY) {
        $self->{_fsm}->Emergency();
    }
    elsif ($route == NYC_TEMP) {
        $self->{_fsm}->NYCTemp();
    }
    elsif ($route == TIME) {
        $self->{_fsm}->Time();
    }
    elsif ($route == DEPOSIT_MONEY) {
        $self->{_fsm}->DepositMoney();
    }
    elsif ($route == LINE_BUSY) {
        $self->{_fsm}->LineBusy();
    }
    elsif ($route == INVALID_NUMBER) {
        $self->{_fsm}->InvalidNumber();
    }
}

package main;

# Display the "telephone" user interface and run until
# the user quits the window.
my $tel = new Telephone();
$tel->Start();
MainLoop();

File: examples/ObjC/EX7/Telephone.m
//
// The contents of this file are subject to the Mozilla Public
// License Version 1.1 (the "License"); you may not use this file
// except in compliance with the License. You may obtain a copy
// of the License at http://www.mozilla.org/MPL/
// 
// Software distributed under the License is distributed on an
// "AS IS" basis, WITHOUT WARRANTY OF ANY KIND, either express or
// implied. See the License for the specific language governing
// rights and limitations under the License.
// 
// The Original Code is State Machine Compiler (SMC).
// 
// The Initial Developer of the Original Code is Charles W. Rapp.
// Portions created by Charles W. Rapp are
// Copyright (C) 2007. Charles W. Rapp.
// All Rights Reserved.
//
// Author
//	Chris Liscio
//
// RCS ID
// $Id: Telephone.m,v 1.2 2009/04/11 13:05:37 cwrapp Exp $
//
// CHANGE LOG
// (See bottom of this file)
//

#import "Telephone.h"
#import "Telephone_sm.h"

static NSArray *gSoundNameArray = nil;

@implementation Telephone

+ (void)initialize
{
    if ( !gSoundNameArray ) {
        gSoundNameArray = [NSArray arrayWithObjects: @"0" , @"1" , @"10" , @"11" , @"12" , @"13" , @"14" , @"15" , @"16" , @"17" , @"18" , @"19" , @"2" , @"20" , @"3" , @"30" , @"4" , @"40" , @"5" , @"50" , @"50_cents_please" , @"6" , @"7" , @"8" , @"9" , @"911" , @"AM" , @"PM" , @"and" , @"busy_signal" , @"could_not_be_completed" , @"dialtone" , @"error_signal" , @"exactly" , @"fast_busy_signal" , @"nyctemp" , @"oclock" , @"oh" , @"phone_off_hook" , @"ring" , @"second" , @"seconds" , @"the_number_you_have_dialed" , @"the_time_is" , @"touch_tone_0" , @"touch_tone_1" , @"touch_tone_10" , @"touch_tone_11" , @"touch_tone_2" , @"touch_tone_3" , @"touch_tone_4" , @"touch_tone_5" , @"touch_tone_6" , @"touch_tone_7" , @"touch_tone_8" , @"touch_tone_9", nil];
    }
}

- (NSMutableDictionary*)playingAudioDict
{
    return _playingAudioDict;
}

- (void)setPlayingAudioDict:(NSMutableDictionary*)aValue
{
    NSMutableDictionary* oldPlayingAudioDict = _playingAudioDict;
    _playingAudioDict = [aValue retain];
    [oldPlayingAudioDict release];
}

- (NSSound*)soundNamed:(NSString*)name
{
    NSSound *ret = [[[NSSound alloc] initWithContentsOfFile:[[NSBundle mainBundle] pathForResource:name ofType:@"au"] byReference:YES] autorelease];
    [ret setDelegate:self];
    return ret;
}

- (void)awakeFromNib
{
    _loopDict = [[NSMutableDictionary alloc] init];
    _timerDict = [[NSMutableDictionary alloc] init];
    _fsm = [[TelephoneContext alloc] initWithOwner:self];
    [_fsm setDebugFlag:YES];
    [_fsm enterStartState];
}

//
// UI Events
//
- (IBAction)digitPressed:(id)sender
{
    if ( [[sender title] isEqualToString:@"*"] ) {
        [_fsm Digit:@"10"];
    } else if ( [[sender title] isEqualToString:@"#"] ) {
        [_fsm Digit:@"11"];
    } else {
        [_fsm Digit:[sender title]];        
    }
}

- (IBAction)offHookPressed:(id)sender
{   
    [_fsm OffHook];
}

- (IBAction)onHookPressed:(id)sender
{
    [_fsm OnHook];    
}

//
// Actions
//

// Getting/setting actions
- (void)setType:(int)type
{
    _type = type;
}
- (int)type
{
    return _type;
}
- (void)addDisplay:(NSString*)str
{
    [_displayString appendString:str];
    [_numberDisplay setString:_displayString];
}
- (void)setDisplayString:(NSString*)d
{
    id old = _displayString;
    _displayString = [d mutableCopy];
    [old release];
}
- (NSString*)displayString
{
    return _displayString;
}
- (void)saveExchange:(NSString*)exch
{
    [_exchange appendString:exch];
    [self addDisplay:exch];
}
- (void)setExchange:(NSString*)exch
{
    id old = _exchange;
    _exchange = [exch mutableCopy];
    [old release];    
}
- (NSString*)exchange
{
    return _exchange;
}
- (void)saveAreaCode:(NSString*)ac
{
    [_areaCode appendString:ac];
    [self addDisplay:ac];
}
- (void)setAreaCode:(NSString*)ac
{
    id old = _areaCode;
    _areaCode = [ac mutableCopy];
    [old release];
}
- (NSString*)areaCode
{
    return _areaCode;
}
- (void)saveLocal:(NSString*)loc
{
    [_local appendString:loc];
    [self addDisplay:loc];
}
- (void)setLocal:(NSString*)loc
{
    id old = _local;
    _local = [loc mutableCopy];
    [old release];
}
- (NSString*)local
{
    return _local;
}

// Display actions
- (void)updateClock
{
    NSDateFormatter *df = [[[NSDateFormatter alloc] init] autorelease];
    [_numberDisplay setString:[df stringFromDate:[NSDate date]]];
}
- (void)clearDisplay
{
    [self setLocal:@""];
    [self setExchange:@""];
    [self setAreaCode:@""];
    [self setDisplayString:@""];

    [_numberDisplay setString:@""];
}

// Receiver button actions
- (void)setReceiver:(SEL)command :(NSString*)text
{
    [_receiverButton setTitle:text];
    [_receiverButton setAction:command];
}

// Timer actions
- (void)startClockTimer
{
    
}
- (void)startTimer:(NSString*)timer :(int)duration
{
    NSInvocation *inv = [NSInvocation invocationWithMethodSignature:[TelephoneContext instanceMethodSignatureForSelector:NSSelectorFromString(timer)]];
    [inv setTarget:_fsm];
    [inv setSelector:NSSelectorFromString(timer)];
    NSTimer *aTimer = [NSTimer scheduledTimerWithTimeInterval:(float)duration / 1000.0
                                                   invocation:inv
                                                      repeats:NO];
    
    [_timerDict setObject:aTimer forKey:timer];
}
- (void)stopTimer:(NSString*)timer
{
    NSTimer *aTimer = [_timerDict objectForKey:timer];
    [aTimer invalidate];
    [_timerDict removeObjectForKey:timer];
}
- (void)resetTimer:(NSString*)timer
{
    NSTimer *aTimer = [_timerDict objectForKey:timer];
    [aTimer invalidate];
    NSDate *fireDate = [NSDate date];
    [fireDate addTimeInterval:[aTimer timeInterval]];
    [aTimer setFireDate:fireDate];
    [[NSRunLoop currentRunLoop] addTimer:aTimer forMode:NSDefaultRunLoopMode];
}

// Phone related actions
- (void)routeCall:(int)type :(NSString*)area :(NSString*)exchange :(NSString*)local
{
    int route = 0;
    
    if (type == EMERGENCY)
    {
        route = EMERGENCY;
    } else if (type == LONG_DISTANCE &&
              [area isEqualToString:@"1212"] &&
              [exchange isEqualToString:@"555"] &&
              [local isEqualToString:@"1234"]) {
        route = NYC_TEMP;
    } else if ([exchange isEqualToString:@"555"])  {
        if ([local isEqualToString:@"1212"]) {
            route = TIME;
        } else {
            route = LINE_BUSY;
        }
    } else if (type == LOCAL) {
        route = DEPOSIT_MONEY;
    } else {
        route = INVALID_NUMBER;
    }
    
    [self performSelector:@selector(callRoute:) withObject:[NSNumber numberWithInt:route] afterDelay:0.1];
}

- (void)callRoute:(NSNumber*)route
{
    switch ([route intValue])
    {
        case EMERGENCY:
            [_fsm Emergency];
            break;
            
        case NYC_TEMP:
            [_fsm NYCTemp];
            break;
            
        case TIME:
            [_fsm Time];
            break;
            
        case DEPOSIT_MONEY:
            [_fsm DepositMoney];
            break;
            
        case LINE_BUSY:
            [_fsm LineBusy];
            break;
            
        case INVALID_NUMBER:
            [_fsm InvalidNumber];
            break;
    }
}

// Sound actions
- (void)loop:(NSString*)name
{
    if ( ![_loopDict objectForKey:name] ) {
        NSSound *sound = [self soundNamed:name];
        [sound setDelegate:self];
        [sound play];
        [_loopDict setObject:sound forKey:name];
    }
}

- (void)stopLoop:(NSString*)name
{
    [[_loopDict objectForKey:name] setDelegate:nil];
    [(NSSound*)[_loopDict objectForKey:name] stop];
    [_loopDict removeObjectForKey:name];
}

- (void)playSoundNamed:(NSString*)sound
{
    NSMutableDictionary *playingAudio = [NSMutableDictionary dictionary];
    NSMutableArray *playList = [NSMutableArray array];
    
    [playingAudio setObject:playList forKey:@"playlist"];
    [playingAudio setObject:[NSNumber numberWithInt:0] forKey:@"cur"];
    
    [playList addObject:[self soundNamed:sound]];
    
    [[playList objectAtIndex:0] play];
    
    [self setPlayingAudioDict:playingAudio];
}

- (void)playTT:(NSString*)num
{
    [[self soundNamed:[NSString stringWithFormat:@"touch_tone_%@", num]] play];
}

- (void)playEmergency
{
    [self playSoundNamed:@"911"];
}

- (void)playNYCTemp
{
    [self playSoundNamed:@"nyctemp"];
}

- (void)playDepositMoney
{
    [self playSoundNamed:@"50_cents_please"];
}

- (void)stopPlayback
{
    int index = [[[self playingAudioDict] objectForKey:@"cur"] intValue];
    NSArray *sounds = [[self playingAudioDict] objectForKey:@"playlist"];
    if ( index < [sounds count] ) {
        [(NSSound*)[sounds objectAtIndex:index] stop];        
    }
}

- (void)soundNumber:(int)num toList:(NSMutableArray*)playList withOh:(BOOL)oh;
{
    if ( num < 10 && oh ) {
        [playList addObject:[self soundNamed:@"oh"]];
        [playList addObject:[self soundNamed:[NSString stringWithFormat:@"%d",num]]];
    } else if ( num < 20 ) {
        [playList addObject:[self soundNamed:[NSString stringWithFormat:@"%d",num]]];
    } else {
        int ones = num % 10;
        int tens = num - ones;
        [playList addObject:[self soundNamed:[NSString stringWithFormat:@"%d",tens]]];
        if ( ones > 0 ) {
            [playList addObject:[self soundNamed:[NSString stringWithFormat:@"%d",ones]]];
        }
    }
}

- (void)soundMeridian:(BOOL)isAM toList:(NSMutableArray*)playList
{
    if ( isAM ) {
        [playList addObject:[self soundNamed:@"AM"]];
    } else {
        [playList addObject:[self soundNamed:@"PM"]];
    }    
}

- (void)playTime
{
    NSCalendar *cal = [[NSCalendar alloc] initWithCalendarIdentifier:NSGregorianCalendar];
    NSDateComponents *dc = [cal components:(NSHourCalendarUnit | NSMinuteCalendarUnit | NSSecondCalendarUnit) fromDate:[NSDate date]];
    int hour = [dc hour];
    int minute = [dc minute];
    int seconds = [dc second];
    int isAM = (hour < 12)?1:0;
    hour = (hour<12)?hour:hour-12;
    
    NSMutableDictionary *playingAudio = [NSMutableDictionary dictionary];
    NSMutableArray *playList = [NSMutableArray array];
    
    [playingAudio setObject:playList forKey:@"playlist"];
    [playingAudio setObject:[NSNumber numberWithInt:0] forKey:@"cur"];
    
    [playList addObject:[self soundNamed:@"the_time_is"]];
    
    // Read the hour
    [playList addObject:[self soundNamed:[NSString stringWithFormat:@"%d", hour]]];
    
    if ( minute == 0 && seconds == 0 ) {
        [playList addObject:[self soundNamed:@"oclock"]];
        [self soundMeridian:isAM toList:playList];
        [playList addObject:[self soundNamed:@"exactly"]];
    } else {
        // Read the minute
        [self soundNumber:minute toList:playList withOh:YES];
        [self soundMeridian:isAM toList:playList];
        
        // Read the seconds
        if ( seconds == 0 ) {
            [playList addObject:[self soundNamed:@"exactly"]];
        } else {
            [playList addObject:[self soundNamed:@"and"]];
            [self soundNumber:seconds toList:playList withOh:NO];
            if ( seconds == 1 ) {
                [playList addObject:[self soundNamed:@"second"]];
            } else {
                [playList addObject:[self soundNamed:@"seconds"]];
            }
        }
    }
    
    [[playList objectAtIndex:0] play];
    
    [self setPlayingAudioDict:playingAudio];
}

- (void)soundPhoneNumber:(NSMutableArray*)list
{
    int i;
    // If this is a long distance number, sound out the
    // area code first.
    if ( [self type] == LONG_DISTANCE ) {
        for( i = 0; i < [[self areaCode] length]; i++ ) {
            [list addObject:[self soundNamed:[[self areaCode] substringWithRange:NSMakeRange(i,1)]]];
        }
    }
    // All types have an exchange.
    for( i = 0; i < [[self exchange] length]; i++ ) {
        [list addObject:[self soundNamed:[[self exchange] substringWithRange:NSMakeRange(i,1)]]];
    }
    // Only long distance and local numbers have a local
    // portion.
    if ([self type] == LONG_DISTANCE || [self type] == LOCAL) {
        for( i = 0; i < [[self local] length]; i++ ) {
            [list addObject:[self soundNamed:[[self local] substringWithRange:NSMakeRange(i,1)]]];
        }
    }
}

- (void)playInvalidNumber
{
    NSMutableDictionary *playingAudio = [NSMutableDictionary dictionary];
    NSMutableArray *playList = [NSMutableArray array];
    
    [playingAudio setObject:playList forKey:@"playlist"];
    [playingAudio setObject:[NSNumber numberWithInt:0] forKey:@"cur"];
    
    [playList addObject:[self soundNamed:@"the_number_you_have_dialed"]];
    [self soundPhoneNumber:playList];
    [playList addObject:[self soundNamed:@"could_not_be_completed"]];
    
    [[playList objectAtIndex:0] play];

    [self setPlayingAudioDict:playingAudio];
}

//
// NSSound delegate method
// 

- (BOOL)isLoopingSound:(NSSound*)sound
{
    NSEnumerator *e = [[_loopDict allValues] objectEnumerator];
    NSSound *s = nil;
    while ( s = [e nextObject] ) {
        if ( [s isEqual:sound] ) {
            return YES;
        }
    }
    return NO;
}

- (void)sound:(NSSound*)sound didFinishPlaying:(BOOL)didFinish
{
    if ( [self isLoopingSound:sound] ) {
        [sound play];
    } else {
        int index = [[[self playingAudioDict] objectForKey:@"cur"] intValue];
        NSArray *sounds = [[self playingAudioDict] objectForKey:@"playlist"];
        if ( index < [sounds count] && [sound isEqual:[sounds objectAtIndex:index]] ) {
            index++;
            [[self playingAudioDict] setObject:[NSNumber numberWithInt:index] forKey:@"cur"];
            if ( index < [sounds count] ) {
                [[sounds objectAtIndex:index] play];
            } else {
                [_fsm PlaybackDone];
            }
        }
    }
}
@end

//
// CHANGE LOG
// $Log: Telephone.m,v $
// Revision 1.2  2009/04/11 13:05:37  cwrapp
// Added enterStartState call.
//
// Revision 1.1  2009/03/01 18:20:39  cwrapp
// Preliminary v. 6.0.0 commit.
//
